ostd/src/mm/frame/meta.rs

```diff
@@ -162,7 +167,126 @@ macro_rules! impl_frame_meta_for {
 
 pub use impl_frame_meta_for;
 
+/// The error type for getting the frame from a physical address.
+#[derive(Debug)]
+pub enum GetFrameError {
+    /// The frame is in use.
+    InUse,
+    /// The frame is not in use.
+    Unused,
+    /// The frame is being initialized or destructed.
+    Busy,
+    /// The frame is private to an owner of [`UniqueFrame`].
+    ///
+    /// [`UniqueFrame`]: super::unique::UniqueFrame
+    Unique,
+    /// The provided physical address is out of bound.
+    OutOfBound,
+    /// The provided physical address is not aligned.
+    NotAligned,
+}
+
+/// Get the reference to a metadata slot.
+fn get_slot(paddr: Paddr) -> Result<&'static MetaSlot, GetFrameError> {
+    if paddr % PAGE_SIZE != 0 {
+        return Err(GetFrameError::NotAligned);
+    }
+    if paddr >= super::MAX_PADDR.load(Ordering::Relaxed) as Paddr {
+        return Err(GetFrameError::OutOfBound);
+    }
+
+    let vaddr = mapping::frame_to_meta::<PagingConsts>(paddr);
+    let ptr = vaddr as *mut MetaSlot;
+
+    // SAFETY: `ptr` points to a valid `MetaSlot` that will never be
+    // mutably borrowed, so taking an immutable reference to it is safe.
+    Ok(unsafe { &*ptr })
+}
+
 impl MetaSlot {
+    /// Initializes the metadata slot of a frame assuming it is unused.
+    ///
+    /// If successful, the function returns a pointer to the metadata slot.
+    /// And the slot is initialized with the given metadata.
+    ///
+    /// The resulting reference count held by the returned pointer is
+    /// [`REF_COUNT_UNIQUE`] if `as_unique_ptr` is `true`, otherwise `1`.
+    pub(super) fn get_from_unused<M: AnyFrameMeta>(
+        paddr: Paddr,
+        metadata: M,
+        as_unique_ptr: bool,
+    ) -> Result<*const Self, GetFrameError> {
+        // Checking unsafe preconditions of the `AnyFrameMeta` trait.
+        debug_assert!(size_of::<M>() <= FRAME_METADATA_MAX_SIZE);
+        debug_assert!(align_of::<M>() <= FRAME_METADATA_MAX_ALIGN);
```

Comment:
We should probably move these checks to `write_meta` so that it covers both `get_from_unused` and `repurpose`.